@here/harp-mapview/lib/image/MapViewImageCache.ts

```diff
@@ -70,34 +84,108 @@ export class MapViewImageCache {
      * Add an image and optionally start loading it. Once done, the [[ImageData]] or [[ImageBitmap]]
      * will be stored in the {@link ImageItem}.
      *
+     * Names are unique within a [[MapView]]. URLs are not unique, multiple images with different
+     * names can have the same URL. Still, URLs are are loaded only once.
+     *
      * @param name - Name of image from {@link @here/harp-datasource-protocol#Theme}.
      * @param url - URL of image.
      * @param startLoading - Optional. Pass `true` to start loading the image in the background.
+     * @param isUserImage - Optional flag, defaults to `false`. Only if `true` the item can be
+     * removed from the cache.
      */
     addImage(
         name: string,
         url: string,
-        startLoading = true
+        startLoading = true,
+        isUserImage = false
     ): ImageItem | Promise<ImageItem | undefined> {
-        const imageItem = this.registerImage(name, url, undefined);
+        const imageItem = this.registerImage(name, url, undefined, isUserImage);
         if (startLoading === true) {
             return ImageCache.instance.loadImage(imageItem);
         }
-
         return imageItem;
     }
 
+    /**
+     * Remove the user image with this name from the cache.
+     *
+     * @param name - Name of the image. Images can only removed if they have been added with
+     * isUserImage set tto `true`.
+     * @returns `true` if item has been removed.
+     */
+    removeImage(name: string): boolean {
+        return this.removeImageInternal(name, false);
+    }
+
+    /**
+     * Remove images using the URL from the cache. Images can only removed if they have
+     * been added with isUserImage set to `true`.
+     *
+     * @param url - URL of the image.
+     * @returns `true` if image has been removed. If multiple images are referring to the same
+     * image URL, they are all removed.
+     */
+    removeImageByUrl(url: string): boolean {
+        const names = this.m_url2Name.get(url);
+        if (names !== undefined) {
+            // Gather the images that are not user images.
+            const themeNames: string[] = [];
```

Comment:
instead of gathering the themeNames, couldn't you just save the index and then splice it from the array in this.m_url2Name?

Reply:
A fine idea. But I would like it to stay like that, it is a loop shorter (over the indices), and the impact of the number array over the string array is, with a small amount of items, probably not measurable.